File: sdk/go/provider/openinbox.go
package provider

import (
	"encoding/json"
	"fmt"
	"io"
	"net/url"
	"strings"

	http "github.com/bogdanfinn/fhttp"
)

const openinboxBase = "https://api.openinbox.io/api"

type openinboxCreateResponse struct {
	ID        string `json:"id"`
	Email     string `json:"email"`
	ExpiresAt string `json:"expiresAt"`
	CreatedAt string `json:"createdAt"`
}

type openinboxListResponse struct {
	Emails []map[string]any `json:"emails"`
}

func openinboxJSON(method, u string, out any) error {
	req, err := http.NewRequest(method, u, nil)
	if err != nil {
		return err
	}
	req.Header.Set("Accept", "application/json, text/plain, */*")
	req.Header.Set("Origin", "https://openinbox.io")
	req.Header.Set("Referer", "https://openinbox.io/")
	req.Header.Set("User-Agent", "Mozilla/5.0")
	if method == http.MethodPost {
		req.Header.Set("Content-Type", "application/json")
	}
	resp, err := HTTPClient().Do(req)
	if err != nil {
		return err
	}
	defer resp.Body.Close()
	data, err := io.ReadAll(resp.Body)
	if err != nil {
		return err
	}
	if resp.StatusCode < 200 || resp.StatusCode >= 300 {
		return fmt.Errorf("openinbox: http %d", resp.StatusCode)
	}
	return json.Unmarshal(data, out)
}

func openinboxFlatten(raw map[string]any, recipient string) map[string]any {
	return map[string]any{
		"id":          raw["id"],
		"from":        raw["from"],
		"to":          recipient,
		"subject":     raw["subject"],
		"text":        raw["textBody"],
		"html":        raw["htmlBody"],
		"receivedAt":  raw["receivedAt"],
		"isRead":      raw["isRead"],
		"attachments": []any{},
	}
}

func OpeninboxGenerate() (*CreatedMailbox, error) {
	var data openinboxCreateResponse
	if err := openinboxJSON(http.MethodPost, openinboxBase+"/inbox", &data); err != nil {
		return nil, err
	}
	id := strings.TrimSpace(data.ID)
	email := strings.TrimSpace(data.Email)
	if id == "" || email == "" || !strings.Contains(email, "@") {
		return nil, fmt.Errorf("openinbox: invalid mailbox response")
	}
	return &CreatedMailbox{
		Channel:   "openinbox",
		Email:     email,
		Token:     id,
		ExpiresAt: data.ExpiresAt,
		CreatedAt: data.CreatedAt,
	}, nil
}

func openinboxFetchDetail(messageID string) (map[string]any, error) {
	var data map[string]any
	u := openinboxBase + "/emails/" + url.PathEscape(messageID)
	if err := openinboxJSON(http.MethodGet, u, &data); err != nil {
		return nil, err
	}
	return data, nil
}

func OpeninboxGetEmails(inboxID, email string) ([]NormEmail, error) {
	box := strings.TrimSpace(inboxID)
	address := strings.TrimSpace(email)
	if box == "" {
		return nil, fmt.Errorf("openinbox: empty inbox id")
	}
	if address == "" {
		return nil, fmt.Errorf("openinbox: empty email")
	}
	u := fmt.Sprintf("%s/emails/inbox/%s?page=1&limit=50", openinboxBase, url.PathEscape(box))
	var data openinboxListResponse
	if err := openinboxJSON(http.MethodGet, u, &data); err != nil {
		return nil, err
	}
	out := make([]NormEmail, 0, len(data.Emails))
	for _, row := range data.Emails {
		raw := row
		if id := strings.TrimSpace(fmt.Sprintf("%v", row["id"])); id != "" {
			if detail, err := openinboxFetchDetail(id); err == nil {
				raw = detail
			}
		}
		out = append(out, NormalizeMap(openinboxFlatten(raw, address), address))
	}
	return out, nil
}

File: sdk/npm/src/providers/openinbox.ts
import { InternalEmailInfo, Email, Channel } from '../types';
import { normalizeEmail } from '../normalize';
import { fetchWithTimeout } from '../retry';

const CHANNEL: Channel = 'openinbox';
const API_BASE = 'https://api.openinbox.io/api';

interface CreateResponse {
  id?: string;
  email?: string;
  expiresAt?: string;
  createdAt?: string;
}

interface ListResponse {
  emails?: ListMessage[];
}

interface ListMessage {
  id?: string;
  from?: string;
  subject?: string;
  receivedAt?: string;
  isRead?: boolean;
  hasAttachments?: boolean;
  threadId?: string;
}

interface DetailResponse extends ListMessage {
  textBody?: string;
  htmlBody?: string;
  headers?: unknown;
}

const HEADERS: Record<string, string> = {
  Accept: 'application/json, text/plain, */*',
  Origin: 'https://openinbox.io',
  Referer: 'https://openinbox.io/',
  'User-Agent': 'Mozilla/5.0',
};

async function fetchJSON<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetchWithTimeout(url, {
    ...init,
    headers: {
      ...HEADERS,
      ...(init?.headers || {}),
    },
  });
  if (!response.ok) {
    throw new Error(`openinbox http ${response.status}`);
  }
  return response.json() as Promise<T>;
}

function flatten(raw: DetailResponse | ListMessage, recipient: string): Record<string, unknown> {
  return {
    id: raw.id || '',
    from: raw.from || '',
    to: recipient,
    subject: raw.subject || '',
    text: (raw as DetailResponse).textBody || '',
    html: (raw as DetailResponse).htmlBody || '',
    receivedAt: raw.receivedAt || '',
    isRead: Boolean(raw.isRead),
    attachments: [],
  };
}

export async function generateEmail(): Promise<InternalEmailInfo> {
  const data = await fetchJSON<CreateResponse>(`${API_BASE}/inbox`, {
    method: 'POST',
    headers: { 'Content-Type': 'application/json' },
  });
  const id = String(data.id || '').trim();
  const email = String(data.email || '').trim();
  if (!id || !email || !email.includes('@')) {
    throw new Error('openinbox: invalid mailbox response');
  }
  return {
    channel: CHANNEL,
    email,
    token: id,
    expiresAt: data.expiresAt,
    createdAt: data.createdAt,
  };
}

async function fetchDetail(id: string): Promise<DetailResponse | null> {
  try {
    return await fetchJSON<DetailResponse>(`${API_BASE}/emails/${encodeURIComponent(id)}`);
  } catch {
    return null;
  }
}

export async function getEmails(token: string, email: string): Promise<Email[]> {
  const inboxId = String(token || '').trim();
  const address = String(email || '').trim();
  if (!inboxId) throw new Error('openinbox: empty inbox id');
  if (!address) throw new Error('openinbox: empty email');

  const data = await fetchJSON<ListResponse>(
    `${API_BASE}/emails/inbox/${encodeURIComponent(inboxId)}?page=1&limit=50`,
  );
  const rows = Array.isArray(data.emails) ? data.emails : [];
  const emails: Email[] = [];
  for (const row of rows) {
    const id = String(row.id || '').trim();
    const detail = id ? await fetchDetail(id) : null;
    emails.push(normalizeEmail(flatten(detail || row, address), address));
  }
  return emails;
}
